src/lib.rs

```diff
@@ -121,18 +126,250 @@ impl Invariant {
     }
 
     pub fn change_fee_receiver(&mut self, pool_key: PoolKey, fee_receiver: ActorId) -> Result<(), InvariantError> {
-        if !self.caller_is_admin() {
+        if !self.is_caller_admin() {
             return Err(InvariantError::NotAdmin);
         }
 
         let mut pool = self.pools.get(&pool_key)?;
         pool.fee_receiver = fee_receiver;
         self.pools.update(&pool_key, &pool)?;
-        
+
         Ok(())
     }
 
-    fn caller_is_admin(&self) -> bool {
+    pub async fn create_position(
+        &mut self,
+        pool_key: PoolKey,
+        lower_tick: i32,
+        upper_tick: i32,
+        liquidity_delta: Liquidity,
+        slippage_limit_lower: SqrtPrice,
+        slippage_limit_upper: SqrtPrice,
+    ) -> Result<Position, InvariantError> {
+        let caller = msg::source();
+        let program = exec::program_id();
+        let current_timestamp = exec::block_timestamp();
+        let current_block_number = exec::block_height() as u64;
+
+        // liquidity delta = 0 => return
+        if liquidity_delta == Liquidity::new(0) {
+            return Err(InvariantError::ZeroLiquidity);
+        }
+
+        let mut pool = self.pools.get(&pool_key)?;
+
+        let mut lower_tick = self
+            .ticks
+            .get(pool_key, lower_tick)
+            .cloned()
+            .unwrap_or_else(|_| Self::create_tick(self, pool_key, lower_tick).unwrap());
+
+        let mut upper_tick = self
+            .ticks
+            .get(pool_key, upper_tick)
+            .cloned()
+            .unwrap_or_else(|_| Self::create_tick(self, pool_key, upper_tick).unwrap());
+
+        let undo_ticks_creation = |invariant: &mut Self, remove_lower_tick: bool, lower_tick: Tick, remove_upper_tick: bool, upper_tick: Tick| {
```

Comment:
Perhaps it can be achieved more easily and safely. You can't just revert the transaction, but handle state rollback instead? I know this may be necessary in the case of cross-program calls like token transfers, but here we are operating within a single contract